### PR TITLE
Enhancements

### PR DESCRIPTION
**Pull request by dudeperf3ct**
Thanks for awesome content. I am enjoying it so far.  I was wondering if you could help me understand the following

In the Overcoming the huge batch size section, why is large training size discouraged or not desirable? My naive assumption was higher batch size faster convergence and less training time.

> If you use, say, 1024 accelerators, you’ll have tiny shards per accelerator and a ton of free memory for micro-batch-size
(MBS), let’s say you can fit MBS=32 - you end up with GBS=32k - most likely not what you want. So you either need to deploy Tensor Parallelism which is non-trivial to implement, or often it’s much simpler to deploy
Sequence Parallelism.

Also, can we not use gradient accumulation to control the batch sizes as well?

### CODE REVIEW

File: network/README.md [hunk 1]
```diff
@@ -543,11 +543,11 @@ This is probably one of the most important multi-segment section that you really
...
 
 ### 1-GPU training
 
-Let's start with a much smaller model of say 2B params, to train it you'd need at least [18 bytes per parameter](../training/performance/README.md#anatomy-of-models-memory-usage) in mixed half precision. So `18*2` 36GB of memory just for model weights, optimizer states and gradients. Plus you need additional memory for activations and it'll depend on the batch size and sequence length. But with 80GB A100 GPU we can definitely train this model on a single GPU.
+Let's start with a much smaller model of say 2B params, to train it you'd need at least [18 bytes per parameter](../training/performance/README.md#anatomy-of-models-memory-usage) in mixed half precision which requires 2 bytes to store. So 36GB (`18*2`) of memory just for model weights, optimizer states and gradients. Plus you need additional memory for activations which depends on the batch size and sequence length. The thumb rule is activation memory scales linearly with batch size and quadratically with the sequence length. Taking 80GB A100 GPU for example, we can definitely train this model on a single GPU.
```
Comment (stas00) [inline network/README.md:550]:
```suggestion
Let's start with a much smaller model of say 2B params, to train it you'd need at least [18 bytes per parameter](../training/performance/README.md#anatomy-of-models-memory-usage) in mixed half precision which requires 2 bytes to store. So 36GB (`18*2`) of memory just for model weights, optimizer states and gradients. Plus you need additional memory for activations, which depends on the batch size and sequence length. A rule of thumb is activation memory scales linearly with batch size and quadratically with the sequence length. Taking 80GB A100 GPU for example, we can definitely train this model on a single GPU.
```

**Review by stas00 — APPROVED** [review]
Thank you very much for the improvements, Shubham.

**Comment by stas00** [comment]
To answer your question - A very large BS hits the law of diminishing returns where more data doesn't improve the signal to SGD and mostly wasted. This is in particular wasteful at the beginning of training where often one deploys a slowly growing global BS because when the weights are random even a tiny amount of data is sufficient to push the weights in the right direction.

Mathematically gradient accumulation of 8 and BS=1K is identical to a 8K global batch with no gradient accumulation, because the step is done on the same gradients (minus some small precision loss in grad accumulation). It's just a technical trick to overcome memory limitations. It's used when the BS you can fit is too small and you want a larger global batch size.
